Mathlib/Topology/UniformSpace/CoveringProfinite.lean

```diff
@@ -0,0 +1,115 @@
+`X → V` can be uniformly approximated by locally constant functions.
+-/
+
+open UniformSpace Set
+
+open TopologicalSpace (Opens Clopens IsOpenCover)
+
+open scoped Uniformity Function
+
+namespace ContinuousMap
+
+variable {X Y V : Type*}
+  [TopologicalSpace X] [TotallyDisconnectedSpace X] [T2Space X] [CompactSpace X]
+  [UniformSpace V] (f : C(X, V)) {S : Set (V × V)}
+
+/--
+For any continuous function on `X` and `S ∈ 𝓤 V`, there exists a finite cover of `X` by
+pairwise-disjoint nonempty clopens, on each of which `f` varies within `S`.
+-/
+lemma exists_disjoint_nonempty_clopen_cover_of_mem_uniformity (hS : S ∈ 𝓤 V) :
+    ∃ (n : ℕ) (D : Fin n → Clopens X), (∀ i, D i ≠ ⊥) ∧
+    (∀ i, ∀ y ∈ D i, ∀ z ∈ D i, (f y, f z) ∈ S) ∧ (univ : Set X) ⊆ ⋃ i, ↑(D i) ∧
+    Pairwise (Disjoint on D) := by
+  -- Find a neighbourhood of each point on which `f` varies within `S`
+  have step1 (x) : ∃ U : Opens X, x ∈ U ∧ ∀ y ∈ U, ∀ z ∈ U, (f y, f z) ∈ S := by
+    obtain ⟨R, hR, hR', hRS⟩ := comp_symm_of_uniformity hS
+    obtain ⟨U, hUB, hUo, hUx⟩ := mem_nhds_iff.mp <|  (f.continuousAt x).preimage_mem_nhds
+      <| UniformSpace.ball_mem_nhds _ hR
+    exact ⟨⟨U, hUo⟩, hUx, fun y hy z hz ↦ hRS <| prodMk_mem_compRel (hR' <| hUB hy) (hUB hz)⟩
+  choose U hUx hUS using step1
+  have hUc : IsOpenCover U := by ext x; simpa using ⟨x, hUx x⟩
+  -- Now refine it to a disjoint covering.
+  obtain ⟨n, W, hW₁, hW₂, hW₃⟩ := hUc.exists_finite_nonempty_disjoint_clopen_cover
+  refine ⟨n, W, fun j ↦ (hW₁ j).1, fun j y hy z hz ↦ ?_, hW₂, hW₃⟩
+  exact match (hW₁ j).2 with | ⟨i, hi⟩ => hUS i y (hi hy) z (hi hz)
+
+/--
+Any continuous function from a profinite space to a uniform space can be uniformly approximated
+by a function factoring through `Fin n`, for some `n`.
+-/
+lemma exists_fin_comp_of_mem_uniformity (hS : S ∈ 𝓤 V) :
+    ∃ (n : ℕ) (g : X → Fin n) (h : Fin n → V), Continuous g ∧ ∀ x, (f x, h (g x)) ∈ S := by
+  classical
+  obtain ⟨n, E, hEne, hES, hEuniv, hEdis⟩ :=
+    exists_disjoint_nonempty_clopen_cover_of_mem_uniformity f hS
+  have h_uniq (x) : ∃! i, x ∈ E i := by
+    refine match Set.mem_iUnion.mp (hEuniv <| mem_univ x) with
+      | ⟨i, hi⟩ => ⟨i, hi, fun j hj ↦ hEdis.eq ?_⟩
+    simpa [← Clopens.coe_disjoint, not_disjoint_iff] using ⟨x, hj, hi⟩
+  choose g hg hg' using h_uniq -- for each `x`, `g x` is the unique `i` such that `x ∈ E i`
+  have h_ex (i) : ∃ x, x ∈ E i := by
+    simpa [← SetLike.coe_set_eq, ← nonempty_iff_ne_empty] using hEne i
+  choose r hr using h_ex -- for each `i`, choose an `r i ∈ E i`
+  refine ⟨n, g, f ∘ r, continuous_discrete_rng.mpr fun j ↦ ?_, fun x ↦ (hES _) _ (hg _) _ (hr _)⟩
+  convert (E j).isOpen
+  exact Set.ext fun x ↦ ⟨fun hj ↦ hj ▸ hg x, fun hx ↦ (hg' _ _ hx).symm⟩
+
+/--
+If `f` is a continuous map from a profinite space to a uniform space with a group structure, then
+we can uniformly approximate `f` by finite products of indicator functions of clopen sets.
```

Comment:
Although it is clear from the statement, you can insist in the doc that no compatibility is required between the uniformity and the group operations. On the other hand, this requires commutativity, which should be mentioned. Also, I simply changed `CommGroup` to `CommMonoid` and it works. This brings me somewhat back to my initial question about `I, Finite I` vs `Fin n`: because commutativity is required precisely because `I` might not be ordered, while since `Fin n` is ordered one can make a sense of `∏ n, f n` even if the target is non-commutative (this is not in Mathlib, AFAIK, and it is really not interesting: I am just advocating for a more general indexing set/type).

Reply:
OK, I agree with you about the commutativity business, but can you spell out explicitely that no compatibility is required in the doc?